src/preamble.js

```diff
@@ -512,17 +512,19 @@ function stringToAscii(str, outPtr) {
 }
 {{{ maybeExport('stringToAscii') }}}
 
-// Given a pointer 'ptr' to a null-terminated UTF8-encoded string in the given array that contains uint8 values, returns
-// a copy of that string as a Javascript String object.
+// Given a pointer 'ptr' to an UTF8-encoded string (optionally null terminated) in the given array that contains uint8 values, 
+// returns a copy of that string as a Javascript String object.
+// len parameter (optional): bytewise length in case string is not null terminated
 
-function UTF8ArrayToString(u8Array, idx) {
+function UTF8ArrayToString(u8Array, idx, len) {
   var u0, u1, u2, u3, u4, u5;
 
   var str = '';
+  var lastIndex = (typeof len !== 'undefined') ? (len + idx + 1) : -1;
   while (1) {
     // For UTF8 byte structure, see http://en.wikipedia.org/wiki/UTF-8#Description and https://www.ietf.org/rfc/rfc2279.txt and https://tools.ietf.org/html/rfc3629
     u0 = u8Array[idx++];
-    if (!u0) return str;
+    if (!u0 || idx == lastIndex) return str;
```

Comment:
This is not correct. The index `idx` can advance between 1 to 6 bytes in an iteration, so a malformed UTF 8 string could cause this check to miss.

Also the variable name `lastIndex` is not that clear, since it is actually two past the last index.

Let's instead have

    var endIdx = (typeof maxBytes === 'undefined') ? Infinity : (idx + maxBytes);
    while(idx < endIdx) {  // +++
        u0 = u8Array[idx++];
        if (!u0) return str;
        ...
    }

and before each `u8Array[idx++]` in the body, check for `if (idx >= endIdx) { return str; }`